src/EditorFeatures/TestUtilities/Completion/AbstractCompletionProviderTests.cs

```diff
@@ -169,64 +198,105 @@ private protected async Task CheckResultsAsync(
-                }
-            }
-            else
-            {
-                if (expectedItemOrNull == null)
-                {
-                    Assert.NotEmpty(items);
+                    if (result.Name == null)
+                    {
+                        Assert.NotEmpty(items);
+                    }
+                    else
+                    {
+                        AssertEx.Any(items, Predicate);
+                    }
                 }
-                else
+
+                bool Predicate(RoslynCompletion.CompletionItem c)
                 {
-                    AssertEx.Any(items, Predicate);
+                    if (!CompareItems(c.DisplayText, result.Name))
+                        return false;
+                    if (!CompareItems(c.DisplayTextSuffix, result.DisplayTextSuffix ?? ""))
+                        return false;
+                    if (!CompareItems(c.DisplayTextPrefix, result.DisplayTextPrefix ?? ""))
+                        return false;
+                    if (!CompareItems(c.InlineDescription, result.InlineDescription ?? ""))
+                        return false;
+                    if (result.ExpectedDescription != null && completionService.GetDescriptionAsync(document, c, options, displayOptions).Result.Text != result.ExpectedDescription)
+                        return false;
+                    if (result.Glyph.HasValue && !c.Tags.SequenceEqual(GlyphTags.GetTags((Glyph)result.Glyph.Value)))
+                        return false;
+                    if (result.MatchPriority.HasValue && c.Rules.MatchPriority != result.MatchPriority.Value)
+                        return false;
+                    if (matchingFilters != null && !FiltersMatch(matchingFilters, c))
+                        return false;
+                    if (flags != null && flags.Value != c.Flags)
+                        return false;
+                    if (result.IsComplexTextEdit is bool textEdit && textEdit != c.IsComplexTextEdit)
+                        return false;
+
+                    return true;
                 }
             }
+        }
+
+        private protected record CompletionTestExpectedResult(
+            string Name,
+            bool IsAbsent,
+            string ExpectedDescription = null,
+            int? Glyph = null,
+            int? MatchPriority = null,
+            string DisplayTextSuffix = null,
+            string DisplayTextPrefix = null,
+            string InlineDescription = null,
+            bool? IsComplexTextEdit = null,
+            SourceCodeKind? SourceCodeKind = null)
+        {
+            public static CompletionTestExpectedResult[] None = CreateGeneralMatchingArray(true);
+            public static CompletionTestExpectedResult[] Any = CreateGeneralMatchingArray(false);
```

Comment:
If these are not intended to be changed (and as far as I can see it is so), please use `ImmutableArray`

Reply:
`VerifyAnyItemExistsAsync` and `VerifyNoItemsExistAsync` use those to pass down to `VerifyExpectedItemsAsync` which takes an array, not `ImmutableArray`. This was designed in the original PR that way, but without any reason against `ImmutableArray` for `VerifyExpectedItemsAsync`.